packages/core/src/toolkit-provider/ToolkitProvider.tsx

```diff
@@ -118,16 +131,50 @@ export function ToolkitProvider({
     (Array.isArray(inheritedThemes) && inheritedThemes.length === 0);
   const density = densityProp ?? inheritedDensity ?? DEFAULT_DENSITY;
   const themeName = getThemeName(themesProp, inheritedThemes);
-  const themes: Theme[] = getTheme(themeName);
+  // We expect theme to be stable
+  const themNameAsString = themeName.toString();
+  const themes: Theme[] = useMemo(
+    () => getTheme(themeName),
+    // if an array is passed to theme inline, themes would be recalculated each time
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [themNameAsString]
```

Comment:
You could avoid the lint break, and benefit from the memoising by grouping together more computed variables in the memo:

```typescript
  const theme = useStaticValue(themesProp);
  const {themeClassnames, themes} = useMemo(() => {
    const themeName = getThemeName(theme, inheritedThemes);
    const themes = getTheme(themeName);
    const themeClassnames = themes.map((theme) => `uitk-${theme.name}`);

    return {
      themes,
      themeClassnames
    }
  }, [theme]);
```

Which is using a simple hook to give meaning to the need for a stable reference:

```typescript
function useStaticValue(input: string | string[]): string[] {
  const stringify = JSON.stringify(input);
  return useMemo(() => JSON.parse(stringify), stringify);
}
```